### DISCUSSION TITLE
xarray revolution

### DISCUSSION THREAD

**Discussion by sbenthall** [econ-ark/HARK#1445] [General]
Continuing discussion from here: 
https://github.com/econ-ark/HARK/pull/1283#issuecomment-1595065727

Perhaps we should use xarray data types everywhere, such as:
 - grids
 - distributions
 - solutions
 - as typed inputs/outputs to equations used in model definitions

@alanlujan91 has modeled this with Labeled distributions and several of the Consumer models already.

I propose (tentatively) that we work these changes throughout the library, rather than reserving them for special 'Labeled' classes.

**Comment by llorracc** [comment]
Agreed.


On Fri, Jun 16, 2023 at 2:55 PM Sebastian Benthall ***@***.***>
wrote:

> Continuing discussion from here:
> #1283 (comment)
> <https://github.com/econ-ark/HARK/pull/1283#issuecomment-1595065727>
>
> Perhaps we should use xarray data types everywhere, such as:
>
>    - grids
>    - distributions
>    - solutions
>    - as typed inputs/outputs to equations used in model definitions
>
> @alanlujan91 <https://github.com/alanlujan91> has modeled this with
> Labeled distributions and several of the Consumer models already.
>
> I propose (tentatively) that we work these changes throughout the library,
> rather than reserving them for special 'Labeled' classes.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/econ-ark/HARK/issues/1289>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKCK7ZOEDLVPPFYEPV65TTXLSTYRANCNFSM6AAAAAAZJTXJRM>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
- Chris Carroll

**Comment by Mv77** [comment]
Just want to add a comment of support. `xarrays` are awesome. Their smart broadcasting enables many useful features that we are already taking advantage of with labeled distributions and in my work on transition matrices.

Only things I would consider would be:
1. They do have a semi-steep learning curve. I was initially frustrated with the coordinates, dimensions, datasets, dataarrays terminology and had to invest in figuring it out. It was absolutely worth it, but if we require too much hands-on with them in model definitions, it might discourage new users.
2. Is it a popular library that we expect to become widespread? My sense is yes, and I see that their community is active. But if we want to put `xarray` at the core of HARK we might want to be extra-confident that the library will be maintained and used. @MridulS might comment on this based on his participation in the py-* conferences?

**Comment by AA-Turner** [comment]
This may be useful: https://hugovk.github.io/top-pypi-packages/ -- ``xarray`` is in the top 1,000 most downloaded packages, and the top 0.2% of packages by downloads.
